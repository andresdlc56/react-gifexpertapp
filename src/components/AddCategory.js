import React, { useState } from 'react';

//Libreria necesaria para la validacion de los props
import PropTypes from 'prop-types';


/* ====================================

    ESTE ES EL SEGUNDO COMPONENTE (HIJO)
    MANEJA EL INPUT 

    ESTE COMPONENTE MODIFICA EL STATE DEL SU COMPONENTE PADRE

======================================= */


export const AddCategory = ({ setCategories }) => {
    
    const [inputValue, setValue] = useState('');


    // DEFINIENDO EVENTO QUE SE VA A DISPARAR CADA VEZ QUE SE TECLEA ALGO EN EL INPUT 
    const handleInputChange = ({ target }) => {

        //CAMBIANDO EL VALOR DEL useState "inputValue" POR EL VALOR DEL TARGET
        setValue(target.value);
    }

    
    // MANEJADOR DEL ENVIO DEL FORMULARIO
    const handleSubmit = (e) => {

        // EVITANDO EL REFRESCAMIENTO DE LA PANTALLA
        e.preventDefault();

        // VALIDACION 
        if(inputValue.trim().length > 2){
            /* DEBIDO A QUE NO TENGO ACCESO A LAS "categories" DIRECTAMENTE DESDE ESTE COMPONENTE POR QUE 
            NO LO PEDI EN LOS PROPS HAGO UNA REFERENCIA DEL MISMO */
            setCategories((cats) => [
                inputValue,
                ...cats
            ]);

            //LIMPIANDO EL VALOR DE INPUT
            setValue('');
        }

        
    }
    
    return (
        <form onSubmit={ handleSubmit }>
            <p>{ inputValue }</p>
            <input 
                type='text'
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>  
    )
}


AddCategory.protoType = {
    setCategories: PropTypes.func.isRequired
}