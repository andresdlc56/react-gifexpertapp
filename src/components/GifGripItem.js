import React from 'react';

import PropTypes from 'prop-types';

export const GifGripItem = ({ title, url }) => {

    return (
        <div className="card animate__animated animate__bounce">
            <img src={ url } alt={ title }/>
            <p>{ title }</p>
        </div>
    )
}

GifGripItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}