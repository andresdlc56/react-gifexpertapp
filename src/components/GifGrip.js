import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGripItem } from './GifGripItem';

//import { getGifs } from '../helpers/getGifs';

import PropTypes from 'prop-types';

export const GifGrip = ({ category }) => {

    /* -------------------------------
        REACT POR NATURALEZA VUELVE A LEER Y EJECUTAR TODO EL COMPONENTE
        AL NOTAR ALGUN CAMBIO EN SI 
    -------------------------------- */


    // IR A BUSCAR GIFS A LA API
    const { data, loading } = useFetchGifs(category);
    

    return (
        <>
            
        
            {
                loading
                    ? <p>LOADING...</p>
                    : 
                    <div>
                        <h3>{ category }</h3>
                        <div className="card-grid">
                            {
                                data.map((img) => (
                                    <GifGripItem 
                                        key={ img.id }
                                        { ...img }
                                    />
                                ))
                            } 
                        </div>
                    </div>
                    
            }

            
        </>
    )
}

GifGrip.protoType = {
    category: PropTypes.string.isRequired
}