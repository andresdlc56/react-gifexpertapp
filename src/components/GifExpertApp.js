import React, { useState } from 'react';
import { AddCategory } from './AddCategory';
import { GifGrip } from './GifGrip';


/* ====================================

    ESTE ES EL COMPONENTE PRINCIAL (PADRE)

======================================= */

export const GifExpertApp = ({ defaultCategories = [] }) => {


    // CREANDO UN useState PARA MANEJAR LAS CATEGORIAS 
    //const [categories, setCategories] = useState(['Dragon Ball']);
    const [categories, setCategories] = useState(defaultCategories);


    return (
        <>
            <h2>GifExpertApp</h2> 

            <AddCategory setCategories={ setCategories }/>

            <hr/> 


            <ol>
                {
                    // BARRIENDO EL ARREGLO DE CATEGORIAS
                    categories.map((category) => ( 
                        <GifGrip 
                            key={ category }
                            category={ category }
                        />
                    ))
                }
            </ol>

          
        </>
    )
}
