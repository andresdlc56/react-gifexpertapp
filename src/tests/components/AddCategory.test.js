import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas en el componente <AddCategory />', () => {

    // REFERENCIA DE FUNCION setCategories(). jest.fn() nos da estadisticas del uso de esta funcion
    const setCategories = jest.fn();

    let wrapper = shallow(<AddCategory setCategories={ setCategories }/>);

    //INSTRUCCION QUE SE VA A EJECUTAR ANTES DE CADA TEST (IMPORTANTISIMO)
    beforeEach(() => {

        // LIMPIA LA SIMULACION DE ALGO EN LOS TESTS
        jest.clearAllMocks();

        //CAPTURANDO EL COMPONENTE A PROBAR
        wrapper = shallow(<AddCategory setCategories={ setCategories }/>);
    });


    test('Deberia mostrar el componente Correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    

    test('Debe cambiar la caja de texto', () => {   
        const input = wrapper.find('input');
        const value = 'Hola';

        // SIMULANDO EVENTO CHANGE EN EL INPUT
        input.simulate('change', { 
            target: { 
                value 
            } 
        });

        const parrafo = wrapper.find('p').text().trim();
        expect(parrafo).toBe(value);
    });
    

    test('No debe postear la informacion con submit', () => {
        
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });

        // NO DEBE LLAMARSE LA FUNCION setCategories() DEBIDO A QUE EL inputValue DEL form ESTA VACIO
        expect(setCategories).not.toHaveBeenCalled();
    });
    

    test('Debe llamar el setCategories() y limpiar la caja de texto', () => {
        
        // CAPTURANDO EL INPUT
        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        // SIMULANDO inputChange
        input.simulate('change', { target: { value } });

        // SIMULANDO EL SUBMIT DEL FORMULARIO
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });

        // VERIFICANDO QUE LA FUNCION setCategories SE EJECUTE
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1); // VERIFICANDO QUE LA FUNCION EVALUADA SE EJECUTE 1 VEZ
        expect(setCategories).toHaveBeenCalledWith( expect.any(Function) ); // VERIFICANDO QUE setCategories llame a otra funcion

        // OBTENIENDO EL VALUE DEL INPUT
        const inputValue = input.prop('value');
        
        // VERIFICANDO QUE EL VALUE DEL INPUT SEA LIMPIADO DESPUES POSTEAR LA INFORMACION 
        expect(inputValue).toBe('');
    });

});