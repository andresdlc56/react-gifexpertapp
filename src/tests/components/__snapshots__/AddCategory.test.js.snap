// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory /> Deberia mostrar el componente Correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
